pkg/auth/providers/goauth/goauth_helper.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		return accounts, err
+	}
+	for _, u := range users {
+		a := Account{
+			SubjectUniqueID: u.Id,
+			Email:           u.PrimaryEmail,
+			PictureURL:      u.ThumbnailPhotoUrl,
+			Type:            userType,
+		}
+		if u.Name != nil {
+			a.Name = u.Name.FullName
+			a.GivenName = u.Name.GivenName
+			a.FamilyName = u.Name.FamilyName
+		}
+		accounts = append(accounts, a)
+	}
+	return accounts, nil
+}
+
+func (g *goauthProvider) searchGroups(searchKey string, config *v3.GoogleOauthConfig) ([]Account, error) {
+	var accounts []Account
+	groupsMap := map[string]*admin.Group{}
+	for _, attr := range []string{"name", "email"} {
+		_, resGroups, err := g.paginateResults(config.Hostname, "", searchKey, attr, groupType, false)
+		if err != nil {
+			return accounts, err
+		}
+		//dedup groups
+		for _, gr := range resGroups {
+			if _, ok := groupsMap[gr.Id]; !ok {
+				groupsMap[gr.Id] = gr
+			}
+		}
+	}
+
+	for _, g := range groupsMap {
+		accounts = append(accounts, Account{SubjectUniqueID: g.Id, Email: g.Email, Name: g.Name, Type: groupType})
+	}
+	return accounts, nil
+}
+
+func (g *goauthProvider) paginateResults(hostedDomain, memberKey, searchKey, searchAttr, accountType string, viewPublic bool) ([]*admin.User, []*admin.Group, error) {
+	var groups []*admin.Group
+	var users []*admin.User
+	switch accountType {
+	case userType:
+		// in case of list users call, using the searchKey itself as query searches all attributes for the search key
+		userListCall := g.adminSvc.Users.List().Domain(hostedDomain).Query(searchKey)
+		if viewPublic {
+			userListCall.ViewType(domainPublicViewType)
+		}
+		for {
+			res, err := userListCall.Do()
+			if err != nil {
+				return users, nil, err
+			}
+			users = append(users, res.Users...)
+			if res.NextPageToken == "" {
+				return users, nil, nil
+			}
+			userListCall.PageToken(res.NextPageToken)
+		}
+	case groupType:
+		groupListCall := g.adminSvc.Groups.List().Domain(hostedDomain)
+		if memberKey != "" {
+			groupListCall.UserKey(memberKey)
+		} else if searchKey != "" {
+			// in case of list groups call, we need to specify search attributes
+			groupListCall.Query(searchAttr + ":" + searchKey + "*")
+		}
+		for {
+			res, err := groupListCall.Do()
+			if err != nil {
+				return nil, groups, err
+			}
+			groups = append(groups, res.Groups...)
+			if res.NextPageToken == "" {
+				return nil, groups, nil
+			}
+			groupListCall.PageToken(res.NextPageToken)
+		}
+	default:
+		return nil, nil, fmt.Errorf("paginateResults: Invalid principal type")
+	}
+	return users, groups, nil
+}
+
+func (g *goauthProvider) directoryServiceFromStoredToken(storedOauthToken string, config *v3.GoogleOauthConfig) error {
+	var oauthToken oauth2.Token
+	if err := json.Unmarshal([]byte(storedOauthToken), &oauthToken); err != nil {
+		return err
+	}
+
+	c, cancel := context.WithTimeout(context.Background(), time.Second*30)
```

Comment:
are we ok with 30 seconds timeout?

Reply:
changed this to use context passed down to Configure